module/zfs/spa_errlog.c

```diff
@@ -128,91 +249,600 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	    DMU_OT_NONE, 0, tx);
+
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa_err_obj);
 	    zap_cursor_retrieve(&zc, &za) == 0;
 	    zap_cursor_advance(&zc)) {
+		if (spa_upgrade_errlog_limit != 0 &&
+		    zc.zc_cd == spa_upgrade_errlog_limit)
+			break;
+
+		name_to_bookmark(za.za_name, &zb);
 
-		if (*count == 0) {
-			zap_cursor_fini(&zc);
-			return (SET_ERROR(ENOMEM));
+		zbookmark_err_phys_t zep;
+		zep.zb_object = zb.zb_object;
+		zep.zb_level = zb.zb_level;
+		zep.zb_blkid = zb.zb_blkid;
+
+		/*
+		 * We cannot use get_head_and_birth_txg() because it will
+		 * acquire the pool config lock, which we already have. In case
+		 * of an error we simply continue.
+		 */
+		uint64_t head_dataset_obj;
+		dsl_pool_t *dp = spa->spa_dsl_pool;
+		dsl_dataset_t *ds;
+		objset_t *os;
+
+		int error = dsl_dataset_hold_obj(dp, zb.zb_objset, FTAG, &ds);
+		if (error != 0)
+			continue;
+
+		head_dataset_obj =
+		    dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj;
+
+		if (dmu_objset_from_ds(ds, &os) != 0) {
+			dsl_dataset_rele(ds, FTAG);
+			continue;
 		}
 
-		name_to_bookmark(za.za_name, &zb);
+		dnode_t *dn;
+		blkptr_t bp;
 
-		if (copyout(&zb, (char *)addr +
-		    (*count - 1) * sizeof (zbookmark_phys_t),
-		    sizeof (zbookmark_phys_t)) != 0) {
-			zap_cursor_fini(&zc);
-			return (SET_ERROR(EFAULT));
+		if (dnode_hold(os, zep.zb_object, FTAG, &dn) != 0) {
+			dsl_dataset_rele(ds, FTAG);
+			continue;
 		}
 
-		*count -= 1;
-	}
+		rw_enter(&dn->dn_struct_rwlock, RW_READER);
+		error = dbuf_dnode_findbp(dn, zep.zb_level, zep.zb_blkid, &bp,
+		    NULL, NULL);
+
+		if (error != 0 || BP_IS_HOLE(&bp)) {
+			rw_exit(&dn->dn_struct_rwlock);
+			dnode_rele(dn, FTAG);
+			dsl_dataset_rele(ds, FTAG);
+			continue;
+		}
+
+		zep.zb_birth = bp.blk_birth;
+		rw_exit(&dn->dn_struct_rwlock);
+		dnode_rele(dn, FTAG);
+		dsl_dataset_rele(ds, FTAG);
+
+		uint64_t err_obj;
+		error = zap_lookup_int_key(spa->spa_meta_objset, *newobj,
+		    head_dataset_obj, &err_obj);
+
+		if (error == ENOENT) {
+			err_obj = zap_create(spa->spa_meta_objset,
+			    DMU_OT_ERROR_LOG, DMU_OT_NONE, 0, tx);
+
+			(void) zap_update_int_key(spa->spa_meta_objset,
+			    *newobj, head_dataset_obj, err_obj, tx);
+		}
 
+		char buf[64];
+		char *name = "";
+		errphys_to_name(&zep, buf, sizeof (buf));
+
+		(void) zap_update(spa->spa_meta_objset, err_obj,
+		    buf, 1, strlen(name) + 1, name, tx);
+	}
 	zap_cursor_fini(&zc);
 
+	VERIFY0(dmu_object_free(spa->spa_meta_objset, spa_err_obj, tx));
+}
+
+void
+spa_upgrade_errlog(spa_t *spa, dmu_tx_t *tx)
+{
+	uint64_t newobj = 0;
+
+	mutex_enter(&spa->spa_errlog_lock);
+	if (spa->spa_errlog_last != 0) {
+		sync_upgrade_errlog(spa, spa->spa_errlog_last, &newobj, tx);
+		spa->spa_errlog_last = newobj;
+	}
+
+	if (spa->spa_errlog_scrub != 0) {
+		sync_upgrade_errlog(spa, spa->spa_errlog_scrub, &newobj, tx);
+		mutex_exit(&spa->spa_errlog_lock);
+		spa->spa_errlog_scrub = newobj;
+	}
+	mutex_exit(&spa->spa_errlog_lock);
```

Comment:
It looks like you're calling `mutex_exit(&spa->spa_errlog_lock)` twice if `(spa->spa_errlog_scrub != 0)`.  Is that expected?

Reply:
Good catch, will fix.

Reply:
I think this was my only change - give me a heads-up when this is in and I'll approve it.

Reply:
Done, thank you!